### PR TITLE
refactor(api): share the cancellable-stream lifecycle across provider clients (#1350)

### LINKED ISSUE #1350: dry-violation: the cancellable-stream lifecycle is hand-rolled four times across the provider clients
## What

Every `ModelApi` implementation has to turn a provider SDK's async iterable into the plugin's `StreamingModelResponse` (`{ complete, cancel }`). All four sites do it by hand, and all four write the *same* seven-step scaffold: a `cancelled` flag → local accumulators → start the SDK stream → `for await (…) { if (cancelled) break; … }` → assemble a `ModelResponse` → `catch (error) { if (cancelled) return <the same partial response>; log; throw }` → return `{ complete, cancel }`.

Only the middle step — how one SDK chunk maps onto text / thoughts / tool calls / usage — is genuinely provider-specific. The lifecycle around it is not, and because it is copied rather than shared, each copy has had to independently rediscover the two non-obvious parts: that the cancelled-in-catch arm must produce a response *identical* to the success arm, and that `cancel()` has to reach the transport rather than only set the flag.

Both of those have already gone wrong once each, in different copies:

- **The identical-response part.** PR #1208 ("dedup identical streaming-response return into a builder") fixed exactly this in `GeminiClient` by extracting a `buildResponse()` closure. `OllamaClient` never got the same treatment and still carries the response literal verbatim twice — `client.ts:178–185` and again at `client.ts:187–195`. `OpenAIClient` independently arrived at the builder shape (`buildResult()`), so two of the four copies have it and one doesn't.
- **The reach-the-transport part.** `GeminiClient.generateStreamingResponse`'s `generateContent` path set the flag and nothing else, so Stop left the request open until the next chunk arrived. Fixed this run in PR #1349 — but the sibling `streamViaInteractions` in the *same file* had been aborting its controller all along, with a comment explaining why, and neither copy could see the other.

This is filed as an issue rather than a PR because the right seam is a design call, not a mechanical edit: the four sites differ in how they hold cancellation state (Ollama assigns `activeStream` after the await and aborts through it; OpenAI and, post-#1349, Gemini own an `AbortController` created before the request; the Interactions path reads a controller off the SDK's stream object), and picking one shape for all of them changes how each client is constructed.

## Evidence

The four sites, all implementing `StreamingModelResponse`:

| # | Site | Response assembly | How `cancel()` reaches the transport |
| --- | --- | --- | --- |
| 1 | `src/api/providers/gemini/client.ts` — `generateStreamingResponse`, `generateContent` path (from L340) | `buildResponse()` closure (#1208) | `AbortController` → `params.config.abortSignal` (added in #1349) |
| 2 | `src/api/providers/gemini/client.ts` — `streamViaInteractions` (from L192) | `InteractionStreamAccumulator.finalize()` | `activeStream?.controller?.abort()`, assigned post-await |
| 3 | `src/api/providers/ollama/client.ts` — `generateStreamingResponse` (L94–213) | **literal duplicated** at L178–185 and L187–195 | `activeStream?.abort()`, assigned post-await, plus an `if (cancelled) stream.abort()` re-check after each await |
| 4 | `src/api/providers/openai/client.ts` — `generateStreamingResponse` (from L138) | `buildResult()` closure | `AbortController` threaded into the request options |

Site 3's post-await assignment is the reason it needs the `if (cancelled) stream.abort()` re-check that sites 1 and 4 don't: the reference doesn't exist yet when `cancel()` may first fire. The comment at `openai/client.ts:150–153` calls this out explicitly as the race the signal-first shape avoids — one client documents the trade-off, and the client that actually has the race doesn't.

Site 3 also drops `promptEvalCount`/`evalCount` into `toUsageMetadata` twice, once per arm, so a future change to usage assembly has two places to land in that file alone.

## Proposed unit of work

- Add a leaf helper — `runCancellableStream` in `src/api/utils/` (a leaf both the clients and nothing else imports, per the `#1155` acyclic-graph rule) — owning: the `cancelled` flag, the `AbortController`, the `for await` + `break`, the `catch` arm that returns the partial on cancel and rethrows otherwise, and the `{ complete, cancel }` return.
- Give it two provider-supplied callbacks: `start(signal) => Promise<AsyncIterable<Chunk>>` and `onChunk(chunk) => void`, plus a `finalize() => ModelResponse` that both the success and cancelled arms call. `finalize` being the single exit is what makes the #1208 class of bug unrepresentable rather than merely fixed.
- Standardise on the signal-first shape (site 1 / site 4): the controller exists before `start` is called, so `cancel()` is safe at any point and the post-await re-check in site 3 disappears. Site 2 is the one that may need an escape hatch — it aborts a controller the SDK hands back rather than one it supplies — so give the helper an optional `onCancel()` for transports that can't take a signal.
- Migrate the four sites one at a time, each with the existing per-client streaming tests kept green. `test/api/providers/openai/client.test.ts:728` and the two cases added in #1349 are the pattern for asserting the signal actually reaches the SDK; Ollama has no equivalent assertion today and should get one as part of its migration.
- Prefer keeping the per-provider chunk decoding exactly where it is. The goal is to shrink each `generateStreamingResponse` to "decode my chunks", not to unify the decoders.

## Out of scope

- `RetryDecorator.generateStreamingResponse` and its `attemptStream` retry loop — that is #1337, a separate duplication one layer up. This issue is strictly the provider clients; whichever way #1337 resolves, these four still each own a stream lifecycle.
- Changing streaming semantics: what a cancelled stream resolves to, whether partial text is retained, chunk-emission timing, or the `StreamCallback` shape. This is a pure restructuring — every client must behave identically before and after.
- The `InteractionStreamAccumulator` and the per-provider chunk decoders. They are the legitimately provider-specific part.
- The non-streaming `generateModelResponse` paths, which have no cancellation and no shared shape worth extracting.

---

_Filed by the `audit-architecture` skill. If this isn't worth doing, close as **Not planned** / `wontfix` — the skill checks closed-not-planned and won't refile._

### PR DESCRIPTION
**Pull request by allenhutchison**
## Summary

Closes #1350

All four provider streaming sites now share one cancellable-stream lifecycle, per the approved plan (the `#1349`-first sequencing note is satisfied — that PR is merged, and its shape is the template).

## The helper — `src/api/utils/cancellable-stream.ts` (new)

`runCancellableStream<TChunk>({ start, onChunk, finalize, onCancel?, onError? }): StreamingModelResponse` owns the whole scaffold every client hand-rolled:

- **Signal-first cancellation.** The `AbortController` exists *before* `start(signal)` runs, so `cancel()` is safe at any point — before, during, or after the awaits. The post-await `if (cancelled) stream.abort()` re-check that Ollama needed disappears for signal-capable transports.
- **`finalize()` is the single exit for both arms.** Success, cancelled mid-iteration, and cancelled-in-catch all return `finalize()` — the #1208 class of bug (two response literals drifting apart) becomes unrepresentable, not merely fixed.
- **`onCancel` escape hatch** for transports that can't take a signal: fired synchronously inside `cancel()`, after the signal aborts, best-effort (throw swallowed). Gemini's SDK-supplied controller and Ollama's `stream.abort()` both ride it.
- **`onError` hook** so each client keeps its own logging; non-cancel errors rethrow.
- Leaf module per the #1155 acyclic rule: provider clients import it, nothing imports back.

Per-provider chunk decoding stays exactly where it was — the goal was to shrink each `generateStreamingResponse` to "decode my chunks", not to unify decoders. `InteractionStreamAccumulator`, `accumulateStreamChunk`, and the Ollama per-path decoders are untouched.

## The four migrations, one commit each

| Site | What changed | Net |
| --- | --- | --- |
| 1. Gemini `generateContent` path | **Already gone** — post-#1508, `generateStreamingResponse` delegates to `streamViaInteractions` only; the issue's site 1 no longer exists | n/a |
| 2. Gemini `streamViaInteractions` | Accumulator stays; SDK-supplied controller reached via `onCancel`; assigned-after-await reference lives on the instance, race-safe because the signal aborts synchronously first | −43/+39 |
| 3. Ollama `generateStreamingResponse` | The **duplicated response literal** (verbatim at both arms, each with its own `toUsageMetadata` call) collapses into one `finalize()`; generate/chat decoders stay per-path; the transport race closes via the signal + `onCancel`, confined to the one place it exists | −91/+114 incl. new test |
| 4. OpenAI `generateStreamingResponse` | Mostly deletion — it already had the signal-first shape; its trade-off comment (why signal-first beats post-hoc abort) **moves onto the helper**, documenting the rule instead of one client's choice; `buildResult()` becomes `finalize()` | −81/+54 |

## Testing

- **New:** `test/api/utils/cancellable-stream-lifecycle.test.ts` — 8 cases: single-exit on all three arms (success / cancelled mid-iteration from inside `onChunk` / cancelled-in-catch / cancel-before-`start`-resolves), `finalize` called exactly once everywhere; signal abort unblocks a stalled read (the #1349 contract, asserted on the helper's own signal); non-cancel error → `onError` + rethrow; `onCancel` fires after abort and its throw is swallowed.
- **Ollama gap closed** (the plan's named gap): `cancel() before the stream resolves aborts the stream once it arrives` — **red-checked** (fails without the post-`start` `signal.aborted` abort, passes with it). The existing `cancel() aborts the underlying stream` and `omits usageMetadata on cancel` cases stay green unchanged.
- **Regression gate:** every existing per-client streaming test passes **unchanged** — Gemini 65, Ollama 38 (was 37; +1 new), OpenAI 47, including the signal-reaches-the-SDK assertions from `openai/client.test.ts:728` and both #1349 cases.
- Full suite: 195 files / 4210 tests green. `typecheck:test` clean, `build` 0 errors, `format-check` clean, `lint` 0 errors (28 warnings — 3 pre-existing, 2 in files I touched are gone), `lint:cycles` clean (0 cycles with the new module in the graph), `knip` clean.

No docs update — pure internal restructuring, no settings/behavior/public-surface change (checked `docs/`, `README.md`; `AGENTS.md`'s API-layer description stays accurate).

## Out of scope (unchanged from the issue)

- `RetryDecorator.generateStreamingResponse`'s retry loop — #1337, one layer up.
- Streaming semantics (what cancel resolves to, partial retention, emission timing, `StreamCallback` shape) — unchanged; every client behaves identically before and after, proven by the unchanged tests.
- Non-streaming `generateModelResponse` paths.

## Screenshots / Screencast

N/A — no runnable UI surface of its own; the lifecycle is exercised through the per-client streaming suites and the new helper tests, all green.

### Required

- [x] I have read and agree to the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have read and agree to the [AI Policy](../AI_POLICY.md)
- [x] This PR is linked to an approved issue where the approach was discussed with a maintainer (#1350, `auto:planned`; plan approved pending #1349, which has merged)
- [x] All CI checks pass (`npm test`, `npm run build`, `npm run format-check`)
- [x] I have tested this change on Desktop — via the streaming suites driving every migrated path (Stop mid-stream is the real surface; the cancel/abort assertions in each suite pin it)
- [x] I have verified this change does not break Mobile (or includes appropriate platform guards) — no platform-specific code; the Ollama browser SDK path is unchanged apart from routing through the helper
- [x] Documentation has been updated (if applicable) — N/A: no user-facing surface
- [x] I understand that I must address all review comments from CodeRabbit and maintainers, or this PR may be closed

### AI-Generated Code

- [x] This PR includes AI-generated or AI-assisted code
- [x] AI tool(s) used: Claude Code (oh-my-pi harness)
- [x] I have reviewed and understand all AI-generated code in this PR

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved cancellation for streaming responses across Gemini, Ollama, and OpenAI.
  - Cancelling before or during stream startup now reliably prevents delayed responses from being processed.
  - Improved handling of stalled or interrupted streams, including cleaner finalization and error reporting.
  - Preserved streamed text, thoughts, tool calls, and usage details when responses complete normally.
- **Tests**
  - Added coverage for successful streaming, early cancellation, stalled streams, errors, and cancellation edge cases.

### CODE REVIEW

File: test/api/providers/ollama/client.test.ts [hunk 1]
```diff
@@ -321,6 +321,36 @@ describe('OllamaClient', () => {
...
+			);
+			// cancel() while `start` is still awaiting the gate — the reference
+			// does not exist yet.
+			streaming.cancel();
+			resolveStream({
+				[Symbol.asyncIterator]: (async function* () {
+					yield { message: { content: 'late' }, done: false };
+				})(),
+				abort,
+			});
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply test/api/providers/ollama/client.test.ts:347]: Fixed in a5f2dc111cf7b72d404a3701cf4eb4fdaba5f78d — the fixture now passes the async-generator function itself, so a guard regression would consume the late chunk and trip the `markdown === ''` assertion instead of being masked by an unusable iterator.

**Comment by allenhutchison** [comment]
Round 2 summary — both findings verified and fixed in `a5f2dc11`:

**1. Fake stream not actually iterable (inline, `test/api/providers/ollama/client.test.ts`)** — valid. `[Symbol.asyncIterator]` held an *invoked* generator object; `for await` needs the function. The cancel-before-iteration path masked it. Fixed: the fixture now passes the generator function itself, so a guard regression consumes the late chunk and trips the `markdown === ''` assertion.

**2. Use the SDK abort signal + per-call stream reference (`gemini/client.ts`) — valid, with one correction to the proposed fix.**

- Shared `activeInteractionsStream` on the instance: agreed, unsafe for overlapping calls on the same client (nothing serializes `generateStreamingResponse`). Removed — the reference is now a per-call closure.
- Signal into `create`: right diagnosis (a cancel while `create` is pending left the HTTP call running), but the SDK's second argument is `GoogleGenAIRequestOptions`, which extends `RequestInit` — the honored key is `signal` (it flows through the `fetchOptions` merge at the fetch layer and onto the `Request`), not `abortSignal`. The suggested `{ abortSignal: signal }` would be silently ignored. Now passing `create(params, { signal })`, plus the post-`start` `signal.aborted` re-check to abort a returned-but-unread stream.
- New regression test: cancel during create-pending captures the signal at the SDK boundary and asserts it aborted — red-checked (fails without the pass-through).

Pre-push: 195 files / 4211 tests green, typecheck:test / build / format-check / lint (0 errors) / lint:cycles / knip all clean.

File: test/api/providers/ollama/client.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (allenhutchison) [reply test/api/providers/ollama/client.test.ts:347]: Valid — cancel() raced the mock's own invocation. Fixed in 2dd2e51a: the mock resolves a create-started notification and the test awaits it before calling stream.cancel(), so the ordering is deterministic (create begun, gate unresolved, then cancel). Red-check intact: the signal/aborted assertions still fail without the pass-through.

**Comment by allenhutchison** [comment]
Round 3 summary — one finding, fixed in `2dd2e51a`:

- **Test race in the create-pending cancel test** (`test/api/providers/gemini/client.test.ts`): valid. `cancel()` ran synchronously after starting the stream, racing the mock's own invocation — the test could pass without exercising the create-pending window at all. The mock now resolves a `create-started` notification and the test awaits it before cancelling, making the ordering deterministic. The signal/aborted assertions still red-check.

Round-2 state: the fixture-iterability thread was resolved (CodeRabbit confirmed), and the outside-diff major (per-call stream reference + SDK `options.signal`) was fixed in `a5f2dc11` — see the round-2 summary for the one correction (the honored key is `signal`, not `abortSignal`).

Gates: 195 files / 4211 tests, typecheck:test, build, format-check, lint (0 errors) — all clean.
